stdlib/public/Cxx/CxxOptional.swift

```diff
@@ -0,0 +1,43 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+public protocol CxxOptional<Wrapped> {
+  associatedtype Wrapped
+
+  func __convertToBool() -> Bool
+
+  var pointee: Wrapped { get }
+}
+
+extension CxxOptional {
+  public var hasValue: Bool {
+    get {
+      return __convertToBool()
+    }
+  }
+
+  public var value: Wrapped {
+    get {
+      return pointee
+    }
+  }
+}
+
+extension Optional {
+  public init<O: CxxOptional>(fromCxx value: O) where O.Wrapped == Wrapped {
```

Comment:
@lorentey Karoy, could you please take a look to see if the signature of this initializer looks reasonable, specifically the `fromCxx` label?
I couldn't make this initializer unlabeled, since `Swift.Optional` already has an unlabeled initializer that wraps around the passed object, which isn't what we'd like to happen here.

Reply:
It looks okay! Naming is difficult as usual, but `fromCxx` is plausible, and if we figure out a better convention, keeping this one around will not hurt much.